File: apps/web/src/services/standup.ts

import { ForbiddenError, NotFoundError } from './sprint';

export type StandupReviewType = 'comment' | 'approve' | 'request_changes';

export interface SaveStandupInput {
  project_id: string;
  org_id: string;
  sprint_id?: string | null;
  author_id: string;
  date: string;
  done: string | null;
  plan: string | null;
  blockers: string | null;
  plan_story_ids?: string[];
}

/**
 * b47f9b05: ee standup GET이 Supabase-direct라 raw plan_story_ids만 내려 FE가 active-sprint scoped
 * stories fallback→백로그(sprint 밖) 탈락. plan_story_ids를 stories org-scope 조회로 resolve해
 * plan_stories(BE #1731 enrich 동형)로 채워 내린다. cross-board(타 프로젝트 백로그) 노출 보장.
 */
export interface PlanStorySummary {
  id: string;
  title: string;
  status: string;
  priority: string;
  project_id: string;
  sprint_id: string | null;
}

export interface CreateStandupFeedbackInput {
  project_id: string;
  org_id: string;
  standup_entry_id: string;
  feedback_by_id: string;
  review_type?: StandupReviewType;
  feedback_text: string;
}

export interface UpdateStandupFeedbackInput {
  review_type?: StandupReviewType;
  feedback_text?: string;
}

export interface StandupFeedbackRecord {
  id: string;
  org_id: string;
  project_id: string;
  sprint_id: string | null;
  standup_entry_id: string;
  feedback_by_id: string;
  review_type: StandupReviewType;
  feedback_text: string;
  created_at: string;
  updated_at: string;
}

export class StandupService {
  constructor(private readonly db: any) {}

  async getEntries(projectId: string, date: string) {
    const { data, error } = await this.db
      .from('standup_entries')
      .select('*')
      .eq('project_id', projectId)
      .eq('date', date)
      .order('created_at');
    if (error) throw error;
    const entries = (data ?? []) as Array<Record<string, unknown>>;
    const map = await this.resolvePlanStories(entries);
    return entries.map((e) => this.attachPlanStories(e, map));
  }

  async getEntryForUser(projectId: string, authorId: string, date: string) {
    const { data, error } = await this.db
      .from('standup_entries')
      .select('*')
      .eq('project_id', projectId)
      .eq('author_id', authorId)
      .eq('date', date)
      .single();
    if (error && error.code !== 'PGRST116') throw error;
    if (!data) return data;
    const map = await this.resolvePlanStories([data]);
    return this.attachPlanStories(data, map);
  }

  /** plan_story_ids → stories org-scope 조회(cross-board)로 PlanStorySummary 맵. admin client에도 org_id 가드. */
  private async resolvePlanStories(
    entries: Array<Record<string, unknown>>,
  ): Promise<Map<string, PlanStorySummary>> {
    const map = new Map<string, PlanStorySummary>();
    const ids = [...new Set(entries.flatMap((e) => (e.plan_story_ids as string[] | null) ?? []))];
    if (ids.length === 0) return map;
    const orgId = (entries.find((e) => e.org_id)?.org_id as string | undefined) ?? null;
    // ⭐ 불변식(디디): org_id 스코프 ONLY + deleted_at is null. sprint/project/board/status 필터 금지
    // — 좁히면 백로그(sprint 밖·타 보드) 다시 탈락(BE _entries_with_plan_stories 동형).
    let query = this.db
      .from('stories')
      .select('id, title, status, priority, project_id, sprint_id')
      .in('id', ids)
      .is('deleted_at', null);
    if (orgId) query = query.eq('org_id', orgId);
    const { data, error } = await query;
    if (error) throw error;
    for (const s of (data ?? []) as Array<Record<string, unknown>>) {
      map.set(s.id as string, {
        id: s.id as string,
        title: s.title as string,
        status: s.status as string,
        priority: s.priority as string,
        project_id: s.project_id as string,
        sprint_id: (s.sprint_id as string | null) ?? null,
      });
    }
    return map;
  }

  private attachPlanStories<T extends Record<string, unknown>>(
    entry: T,
    map: Map<string, PlanStorySummary>,
  ): T & { plan_stories: PlanStorySummary[] } {
    const ids = (entry.plan_story_ids as string[] | null) ?? [];
    return {
      ...entry,
      plan_stories: ids
        .map((id) => map.get(id))
        .filter((s): s is PlanStorySummary => Boolean(s)),
    };
  }

  async save(input: SaveStandupInput) {
    const { data, error } = await this.db
      .from('standup_entries')
      .upsert({
        project_id: input.project_id,
        org_id: input.org_id,
        sprint_id: input.sprint_id ?? null,
        author_id: input.author_id,
        date: input.date,
        done: input.done,
        plan: input.plan,
        blockers: input.blockers,
        plan_story_ids: input.plan_story_ids ?? [],
      }, { onConflict: 'project_id,author_id,date' })
      .select()
      .single();
    if (error) throw error;
    return data;
  }

  async getMissing(projectId: string, date: string) {
    // deadline 조회: 오늘 날짜면 마감 전 여부 체크
    const today = new Date().toISOString().slice(0, 10);
    if (date === today) {
      const { data: settings } = await this.db
        .from('project_settings')
        .select('standup_deadline')
        .eq('project_id', projectId)
        .maybeSingle();
      const deadline = (settings?.standup_deadline as string | null) ?? '09:00';
      const [dh, dm] = deadline.split(':').map(Number);
      const now = new Date();
      const deadlineTime = new Date(now);
      deadlineTime.setHours(dh, dm, 0, 0);
      if (now < deadlineTime) {
        return { submitted_count: 0, missing: [], deadline_not_reached: true };
      }
    }

    const { data: members } = await this.db
      .from('team_members')
      .select('id, name')
      .eq('project_id', projectId)
      .eq('is_active', true);
    const { data: entries } = await this.db
      .from('standup_entries')
      .select('author_id')
      .eq('project_id', projectId)
      .eq('date', date);
    const submitted = new Set((entries ?? []).map((e) => e.author_id as string));
    const missing = (members ?? []).filter((m) => !submitted.has(m.id as string));
    return { submitted_count: submitted.size, missing: missing.map((m) => ({ id: m.id, name: m.name })) };
  }

  async getHistory(projectId: string, limit = 50) {
    const { data, error } = await this.db
      .from('standup_entries')
      .select('author_id, date, done, plan, blockers, plan_story_ids, org_id')
      .eq('project_id', projectId)
      .order('date', { ascending: false })
      .limit(limit);
    if (error) throw error;
    const entries = (data ?? []) as Array<Record<string, unknown>>;
    const map = await this.resolvePlanStories(entries);
    return entries.map((e) => this.attachPlanStories(e, map));
  }
}

export class StandupFeedbackService {
  constructor(private readonly db: any) {}

  async listByDate(projectId: string, date: string): Promise<StandupFeedbackRecord[]> {
    const { data: entries, error: entriesError } = await this.db
      .from('standup_entries')
      .select('id')
      .eq('project_id', projectId)
      .eq('date', date);

    if (entriesError) throw entriesError;
    if (!entries || entries.length === 0) return [];

    const entryIds = entries.map((entry) => entry.id);
    const { data, error } = await this.db
      .from('standup_feedback')
      .select('*')
      .eq('project_id', projectId)
      .in('standup_entry_id', entryIds)
      .order('created_at');

    if (error) throw error;
    return (data ?? []) as StandupFeedbackRecord[];
  }

  async create(input: CreateStandupFeedbackInput): Promise<StandupFeedbackRecord> {
    const { data: entry, error: entryError } = await this.db
      .from('standup_entries')
      .select('id, org_id, project_id, sprint_id')
      .eq('id', input.standup_entry_id)
      .single();

    if (entryError) {
      if (entryError.code === 'PGRST116') throw new NotFoundError('Standup entry not found');
      if (entryError.code === '42501') throw new ForbiddenError('Permission denied');
      throw entryError;
    }

    if (entry.project_id !== input.project_id || entry.org_id !== input.org_id) {
      throw new ForbiddenError('Permission denied');
    }

    const feedbackText = input.feedback_text.trim();
    if (!feedbackText) throw new Error('feedback_text is required');

    const { data, error } = await this.db
      .from('standup_feedback')
      .insert({
        org_id: entry.org_id,
        project_id: entry.project_id,
        sprint_id: entry.sprint_id ?? null,
        standup_entry_id: input.standup_entry_id,
        feedback_by_id: input.feedback_by_id,
        review_type: input.review_type ?? 'comment',
        feedback_text: feedbackText,
      })
      .select()
      .single();

    if (error) {
      if (error.code === '42501') throw new ForbiddenError('Permission denied');
      throw error;
    }
    return data as StandupFeedbackRecord;
  }

  async update(id: string, input: UpdateStandupFeedbackInput, feedbackById: string): Promise<StandupFeedbackRecord> {
    const sanitized: Record<string, unknown> = {};
    if (input.review_type) sanitized.review_type = input.review_type;
    if (input.feedback_text !== undefined) {
      const feedbackText = input.feedback_text.trim();
      if (!feedbackText) throw new Error('feedback_text is required');
      sanitized.feedback_text = feedbackText;
    }
    if (Object.keys(sanitized).length === 0) throw new Error('No valid fields to update');

    const { data: existing, error: existingError } = await this.db
      .from('standup_feedback')
      .select('id, feedback_by_id')
      .eq('id', id)
      .single();

    if (existingError) {
      if (existingError.code === 'PGRST116') throw new NotFoundError('Standup feedback not found');
      if (existingError.code === '42501') throw new ForbiddenError('Permission denied');
      throw existingError;
    }

    if (!existing || existing.feedback_by_id !== feedbackById) {
      throw new ForbiddenError('Permission denied');
    }

    const { data, error } = await this.db
      .from('standup_feedback')
      .update(sanitized)
      .eq('id', id)
      .eq('feedback_by_id', feedbackById)
      .select()
      .single();

    if (error) {
      if (error.code === 'PGRST116') throw new NotFoundError('Standup feedback not found');
      if (error.code === '42501') throw new ForbiddenError('Permission denied');
      throw error;
    }
    return data as StandupFeedbackRecord;
  }

  async delete(id: string, feedbackById: string): Promise<void> {
    const { data: existing, error: existingError } = await this.db
      .from('standup_feedback')
      .select('id, feedback_by_id')
      .eq('id', id)
      .single();

    if (existingError) {
      if (existingError.code === 'PGRST116') throw new NotFoundError('Standup feedback not found');
      if (existingError.code === '42501') throw new ForbiddenError('Permission denied');
      throw existingError;
    }

    if (!existing || existing.feedback_by_id !== feedbackById) {
      throw new ForbiddenError('Permission denied');
    }

    const { error } = await this.db
      .from('standup_feedback')
      .delete()
      .eq('id', id)
      .eq('feedback_by_id', feedbackById);

    if (error) {
      if (error.code === '42501') throw new ForbiddenError('Permission denied');
      throw error;
    }
  }
}

File: backend/app/schemas/standup.py
import uuid
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, model_validator

from app.schemas.validators import reject_if_all_blank

REVIEW_TYPES = ("comment", "approve", "request_changes")


class StandupUpsert(BaseModel):
    # 1c2be9db: org-level write — project_id 생략 가능. 생략 시 entry 링크 = author 접근
    # 프로젝트 자동(accessible_project_ids_in_org). 명시 시 그 프로젝트 additive 링크(legacy).
    project_id: uuid.UUID | None = None
    org_id: uuid.UUID | None = None
    author_id: uuid.UUID
    date: date
    sprint_id: uuid.UUID | None = None
    done: str | None = None
    plan: str | None = None
    blockers: str | None = None
    plan_story_ids: list[uuid.UUID] = []

    # story #2414(PO 지시, 2026-08-02) — 세 칸이 전부 빈 스탠드업이 저장돼 있었다(실측
    # 2건). done만 빈 것(오늘 시작한 사람)은 정상이라 막지 않는다 — 실측 122건 중 73건이
    # 이 모양이라 그것까지 막으면 과잉차단이다. 셋 다 빌 때만 거부.
    @model_validator(mode="after")
    def _reject_all_blank_content(self) -> "StandupUpsert":
        reject_if_all_blank(done=self.done, plan=self.plan, blockers=self.blockers)
        return self


class StandupSelfUpdate(BaseModel):
    """self-save(PUT) 전용 — author_id는 서버가 인증 유저(resolve_member)에서 도출.

    클라 바디 author_id를 받지 않아 타인 스탠드업 위조를 차단(SID:6a1e8b1d).
    project_id는 바디 수용(생략 가능 — 1c2be9db org-level). (extra 필드는 pydantic 기본 무시)
    """
    project_id: uuid.UUID | None = None
    date: date
    sprint_id: uuid.UUID | None = None
    done: str | None = None
    plan: str | None = None
    blockers: str | None = None
    plan_story_ids: list[uuid.UUID] = []

    # story #2414 — POST(StandupUpsert)와 동일 규칙, PUT도 같은 결함을 낼 수 있어 동형 적용.
    @model_validator(mode="after")
    def _reject_all_blank_content(self) -> "StandupSelfUpdate":
        reject_if_all_blank(done=self.done, plan=self.plan, blockers=self.blockers)
        return self


class StandupEntryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    project_id: uuid.UUID | None = None  # 1c2be9db: org-level 엔트리는 origin project_id 가 NULL 일 수 있음
    org_id: uuid.UUID
    sprint_id: uuid.UUID | None = None
    author_id: uuid.UUID
    date: date
    done: str | None = None
    plan: str | None = None
    blockers: str | None = None
    plan_story_ids: list[uuid.UUID]
    # a9e67531: plan_story_ids 의 org-scope resolve 요약(id+title+status). FE 가 active-sprint stories 배열에
    # 의존하지 않고 cross-board plan story 도 렌더(미노출 버그 근본 fix). plan_story_ids 는 하위호환 유지.
    plan_stories: list["PlanStorySummary"] = []
    created_at: datetime
    updated_at: datetime


class PlanStorySummary(BaseModel):
    id: uuid.UUID
    title: str
    status: str
    priority: str | None = None
    project_id: uuid.UUID | None = None
    sprint_id: uuid.UUID | None = None


class FeedbackCreate(BaseModel):
    org_id: uuid.UUID
    project_id: uuid.UUID
    sprint_id: uuid.UUID | None = None
    feedback_by_id: uuid.UUID
    review_type: str = "comment"
    feedback_text: str


class FeedbackResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    project_id: uuid.UUID
    org_id: uuid.UUID
    sprint_id: uuid.UUID | None = None
    standup_entry_id: uuid.UUID
    feedback_by_id: uuid.UUID
    review_type: str
    feedback_text: str
    created_at: datetime
    updated_at: datetime
